server/drafty/drafty.go

```diff
@@ -22,368 +47,572 @@ type span struct {
+func Preview(content interface{}, length int) (string, error) {
+	doc, err := decodeAsDrafty(content)
 	if err != nil {
-		return err
+		return "", err
+	}
+	if doc == nil {
+		return "", nil
 	}
 
-	s.end, err = intFromNumeric(m["len"])
+	tree, err := toTree(doc)
 	if err != nil {
-		return err
+		return "", err
 	}
-	if s.end < 0 {
-		return errInvalidContent
+	if tree == nil {
+		return "", nil
 	}
-	s.end += s.at
 
-	if s.tp == "" {
-		s.key, err = intFromNumeric(m["key"])
-		if err != nil {
-			return err
-		}
-		if s.key < 0 {
-			return errInvalidContent
-		}
+	state := previewState{
+		drafty: &document{
+			Fmt: make([]style, 0, len(doc.Fmt)),
+			Ent: make([]entity, 0, len(doc.Ent)),
+		},
+		maxLength: length,
+		keymap:    make(map[int]int),
 	}
 
-	return nil
+	if err = previewFormatter(tree, &state); err != nil {
+		return "", err
+	}
+
+	state.drafty.Txt = string(state.drafty.txt)
+	data, err := json.Marshal(state.drafty)
+	return string(data), err
 }
 
-func intFromNumeric(num interface{}) (int, error) {
-	if num == nil {
-		return 0, nil
+type plainTextState struct {
+	txt string
+}
+
+// PlainText converts drafty document to plain text with some basic markdown-like formatting.
+// Deprecated. Use Preview for new development.
+func PlainText(content interface{}) (string, error) {
+	doc, err := decodeAsDrafty(content)
+	if err != nil {
+		return "", err
 	}
-	switch i := num.(type) {
-	case int:
-		return i, nil
-	case int16:
-		return int(i), nil
-	case int32:
-		return int(i), nil
-	case int64:
-		return int(i), nil
-	case float32:
-		return int(i), nil
-	case float64:
-		return int(i), nil
-	default:
-		return 0, errInvalidContent
+	if doc == nil {
+		return "", nil
 	}
-}
 
-func (s *span) toMap() map[string]interface{} {
-	out := make(map[string]interface{})
+	tree, err := toTree(doc)
+	if err != nil {
+		return "", err
+	}
 
-	if s.tp != "" {
-		out["tp"] = s.tp
-	} else {
-		out["key"] = s.key
+	state := plainTextState{}
+
+	err = plainTextFormatter(tree, &state)
+	if err != nil {
+		return "", err
 	}
-	if s.at != 0 {
-		out["at"] = s.at
+
+	return strings.TrimSpace(string(state.txt)), nil
+}
+
+// styleToSpan converts Drafty style to internal representation.
+func (s *span) styleToSpan(in *style) error {
+	s.tp = in.Tp
+	s.at = in.At
+	s.end = in.Length
+	if s.end < 0 {
+		return errInvalidContent
 	}
-	if s.end != s.at {
-		out["len"] = s.end - s.at
+	s.end += s.at
+
+	if s.tp == "" {
+		s.key = in.Key
+		if s.key < 0 {
+			return errInvalidContent
+		}
 	}
 
-	return out
+	return nil
 }
 
 type spanfmt struct {
 	dec    string
 	isVoid bool
 }
 
+// Plain text formatting of the Drafty tags.
 var tags = map[string]spanfmt{
 	"ST": {"*", false},
 	"EM": {"_", false},
 	"DL": {"~", false},
-	"CO": {"", false},
+	"CO": {"`", false},
 	"BR": {"\n", true},
 	"LN": {"", false},
 	"MN": {"", false},
 	"HT": {"", false},
-	"IM": {"", true},
+	"IM": {"", false},
 	"EX": {"", true},
+	"QQ": {"", false},
```

Comment:
Can you add `IC` here as well?

Reply:
Actually, only non-blank tags are needed here. I added `IC` handling further down in the code + test.